sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py

```diff
@@ -157,49 +159,64 @@ def _create_handler(self, auth: "JWTTokenAsync") -> None:
         await self._do_retryable_operation(self._open, operation_need_param=False)
 
     def _message_received(self, message: uamqp.Message) -> None:
-        self._event_queue.put(message)
+        self._message_buffer.appendleft(message)
 
-    async def receive(self) -> None:
-        retried_times = 0
-        last_exception = None
+    def _next_message_in_buffer(self):
+        # pylint:disable=protected-access
+        message = self._message_buffer.pop()
+        event_data = EventData._from_message(message)
+        trace_link_message(event_data)
+        self._last_received_event = event_data
+        return event_data
+
+    async def receive(self, batch=False, max_batch_size=300, max_wait_time=None) -> None:
         max_retries = (
             self._client._config.max_retries  # pylint:disable=protected-access
         )
-
-        while retried_times <= max_retries:
-            try:
-                await self._open()
-                await cast(ReceiveClientAsync, self._handler).do_work_async()
-                break
-            except asyncio.CancelledError:  # pylint: disable=try-except-raise
-                raise
-            except Exception as exception:  # pylint: disable=broad-except
-                if (
-                    isinstance(exception, uamqp.errors.LinkDetach)
-                    and exception.condition == uamqp.constants.ErrorCodes.LinkStolen  # pylint: disable=no-member
-                ):
-                    raise await self._handle_exception(exception)
-                if not self.running:  # exit by close
-                    return
-                if self._last_received_event:
-                    self._offset = self._last_received_event.offset
-                last_exception = await self._handle_exception(exception)
-                retried_times += 1
-                if retried_times > max_retries:
-                    _LOGGER.info(
-                        "%r operation has exhausted retry. Last exception: %r.",
-                        self._name,
-                        last_exception,
-                    )
-                    raise last_exception
-
-        while not self._event_queue.empty():
-            message = self._event_queue.get()
-            event_data = EventData._from_message(  # pylint:disable=protected-access
-                message
-            )
-            self._last_received_event = event_data
-            trace_link_message(event_data)
-            await self._on_event_received(event_data)
-            self._event_queue.task_done()
-        return
+        has_not_fetched_once = True  # ensure one trip when max_wait_time is very small
+        deadline = time.time() + (max_wait_time or 0)  # max_wait_time can be None
+        while len(self._message_buffer) < max_batch_size and \
+                (time.time() < deadline or has_not_fetched_once):
```

Comment:
It seems that we're always checking deadline even if users have set no `max_wait_time`?

Also - I didn't see the `has_not_fetched_once` mechanism in the sync version - it's not needed there?

Reply:
I don't think this currently accommodates the "return events immediately if no wait time set" - am I missing something?

Reply:
1. sync and async have different implementation. Each partition receive is running in a Task for async. So it can repeatedly run for events for a while and other partition receiving won't get blocked. Sync is different because it's single-threaded for all partitions. You may want to look at their implementation separately. If sync is multi-threaded, their implementation should be the same.
2. when max_wait_time is not set or it's 0 the deadline is the current time. so time.time() < deadline is False. Then has_not_fetch_once kicks in to ensure to fetch once if len(self._message_buffer) < max_batch_size. The while loop then end.